applications/StructuralMechanicsApplication/custom_utilities/large_displacement_variables.cpp

```diff
@@ -0,0 +1,328 @@
+// KRATOS  ___|  |                   |                   |
+//       \___ \  __|  __| |   |  __| __| |   |  __| _` | |
+//             | |   |    |   | (    |   |   | |   (   | |
+//       _____/ \__|_|   \__,_|\___|\__|\__,_|_|  \__,_|_| MECHANICS
+//
+//  License:		 BSD License
+//					 license: structural_mechanics_application/license.txt
+//
+//  Main authors:    
+//
+
+// System includes
+
+// External includes
+
+// Project includes
+#include "custom_utilities/large_displacement_variables.h"
+#include "utilities/matrix_vector_utilities.h"
+#include "utilities/math_utils.h"
+#include "utilities/geometry_utilities.h"
+#include "custom_utilities/structural_mechanics_math_utilities.hpp"
+
+namespace Kratos
+{
+LargeDisplacementDifferentialVariables::LargeDisplacementDifferentialVariables(GeometryType const& rGeom)
+    : mrGeom(rGeom)
+{
+}
+
+const Matrix& LargeDisplacementDifferentialVariables::F(std::size_t IntegrationIndex, bool IsAxisymmetric)
+{
+    KRATOS_TRY;
+    if (IntegrationIndex != mIntegrationIndexF)
+        CalculateF(IntegrationIndex, IsAxisymmetric);
+    return mF;
+    KRATOS_CATCH("");
+}
+
+double LargeDisplacementDifferentialVariables::DetF(std::size_t IntegrationIndex, bool IsAxisymmetric)
+{
+    KRATOS_TRY;
+    if (IntegrationIndex != mIntegrationIndexDetF)
+        CalculateDetF(IntegrationIndex, IsAxisymmetric);
+    return mDetF;
+    KRATOS_CATCH("");
+}
+
+const Matrix& LargeDisplacementDifferentialVariables::DN_DX0(std::size_t IntegrationIndex)
+{
+    KRATOS_TRY;
+    if (IntegrationIndex != mIntegrationIndexDN_DX0)
+        CalculateDN_DX0(IntegrationIndex);
+    return mDN_DX0;
+    KRATOS_CATCH("");
+}
+
+double LargeDisplacementDifferentialVariables::DetJ0(std::size_t IntegrationIndex)
+{
+    KRATOS_TRY;
+    if (IntegrationIndex != mInternalIntegrationIndex)
+        RecalculateInternals(IntegrationIndex);
+    KRATOS_ERROR_IF(mDetJ0 < 0.0) << " INVERTED. DETJ0: " << mDetJ0 << " at "
+                                  << mrGeom.Center() << '.' << std::endl;
+    return mDetJ0;
+    KRATOS_CATCH("");
+}
+
+void LargeDisplacementDifferentialVariables::RecalculateInternals(std::size_t IntegrationIndex)
```

Comment:
I would leave as public only this. I think it is useful that F and all the others are computed all at once, so that they are in sync...this is so in the hypothesis you want to store their values.

The point is that if you don't do this you may well have two of them computed at different moments and yielding different values (for example corresponding to a different IntegrationIndex) this would be awful to debug...